src/vm/fieldmarshaler.cpp

```diff
@@ -3364,14 +3364,22 @@ VOID FieldMarshaler_FixedStringAnsi::UpdateNativeImpl(OBJECTREF* pCLRValue, LPVO
             nc = m_numchar - 1;
 
         int cbwritten = InternalWideToAnsi(pString->GetBuffer(),
-                                           nc,
-                                            (CHAR*)pNativeValue,
-                                           m_numchar,
-                                           m_BestFitMap,
-                                           m_ThrowOnUnmappableChar);
+            nc,
+            (CHAR*)pNativeValue,
+            m_numchar,
+            m_BestFitMap,
+            m_ThrowOnUnmappableChar);
+
+        // Handle the case where SizeConst == Number of bytes.For single byte chars 
+        // this will never be the case since nc >= m_numchar check will truncate the last 
+        // character, but for multibyte chars nc>= m_numchar check won't truncate since GetStringLength
+        // gives number of characters but not the actual number of bytes. For such cases need to make 
+        // sure that we dont write one past the buffer.
+        if (cbwritten == m_numchar)
+            --cbwritten;
```

Comment:
This means we'll potentially end up partial characters (for example, half of Chinese) if you overwrite the 2nd half with 0. Why not simply pass m_numchar - 1?